File: src/main/java/ru/ioleksiv/telegram/bot/api/annotations/filter/telegram/SuccessfulPaymentFilter.java
package ru.ioleksiv.telegram.bot.api.annotations.filter.telegram;

import ru.ioleksiv.telegram.bot.api.annotations.filter.primitive.IntegerFilter;
import ru.ioleksiv.telegram.bot.api.annotations.filter.primitive.NotNullFilter;
import ru.ioleksiv.telegram.bot.api.annotations.filter.primitive.StringFilter;
import ru.ioleksiv.telegram.bot.api.model.annotation.AnnotationState;
import ru.ioleksiv.telegram.bot.api.model.annotation.CustomValidator;
import ru.ioleksiv.telegram.bot.api.model.annotation.stub.StubSuccessfulPaymentValidator;
import ru.ioleksiv.telegram.bot.api.model.objects.payments.SuccessfulPayment;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
@Documented
public @interface SuccessfulPaymentFilter {

    AnnotationState status() default AnnotationState.ON;

    String[] validator() default {};

    StringFilter currency() default @StringFilter(status = AnnotationState.OFF);

    IntegerFilter totalAmount() default @IntegerFilter(status = AnnotationState.OFF);

    StringFilter invoicePayload() default @StringFilter(status = AnnotationState.OFF);

    NotNullFilter shippingOptionId() default @NotNullFilter(status = AnnotationState.OFF);

    OrderInfoFilter orderInfo() default @OrderInfoFilter(status = AnnotationState.OFF);

    NotNullFilter telegramPaymentChargeId() default @NotNullFilter(status = AnnotationState.OFF);

    NotNullFilter providerPaymentChargeId() default @NotNullFilter(status = AnnotationState.OFF);

}

File: core/src/main/kotlin/io/github/oleksivio/tl/kbot/core/annotations/filter/composite/SuccessfulPaymentFilter.kt
package io.github.oleksivio.tl.kbot.core.annotations.filter.composite

import io.github.oleksivio.tl.kbot.core.annotations.filter.primitive.IntegerFilter
import io.github.oleksivio.tl.kbot.core.annotations.filter.primitive.NotNullFilter
import io.github.oleksivio.tl.kbot.core.annotations.filter.primitive.StringFilter
import io.github.oleksivio.tl.kbot.core.model.annotation.AnnotationState

@Retention
@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
@MustBeDocumented
annotation class SuccessfulPaymentFilter(
    val status: AnnotationState = AnnotationState.ON,
    val validator: Array<String> = [],
    val currency: StringFilter = StringFilter(
        status = AnnotationState.OFF
    ),
    val totalAmount: IntegerFilter = IntegerFilter(
        status = AnnotationState.OFF
    ),
    val invoicePayload: StringFilter = StringFilter(
        status = AnnotationState.OFF
    ),
    val shippingOptionId: NotNullFilter = NotNullFilter(
        status = AnnotationState.OFF
    ),
    val orderInfo: OrderInfoFilter = OrderInfoFilter(
        status = AnnotationState.OFF
    ),
    val telegramPaymentChargeId: NotNullFilter = NotNullFilter(
        status = AnnotationState.OFF
    ),
    val providerPaymentChargeId: NotNullFilter = NotNullFilter(
        status = AnnotationState.OFF
    )
)
